pacman/utilities/algorithm_utilities/routing_info_allocator_utilities.py

```diff
@@ -26,39 +32,542 @@ def get_edge_groups(machine_graph, traffic_type):
+    partitions which share the same key, constraint group)
+    :rtype: None
+    """
+
+    # get types of constraints from this partition
+    (is_continuous, is_fixed_mask, is_fixed_key, is_flexi_field,
+     is_fixed_field, is_shared_key) = _check_types_of_constraints(
+        partition.constraints)
+
+    # this partition already in a share key somewhere, merge in
+    if partition in partition_mappings:
+
+        # locate previous referral
+        tracked_set, group_currently_resides_in = partition_mappings[partition]
+
+        # locate real group
+        real_group = _find_group(
+            fixed_key_groups=fixed_key_groups,
+            fixed_mask_groups=fixed_mask_groups,
+            fixed_field_groups=fixed_field_groups,
+            flexi_field_groups=flexi_field_groups,
+            continuous_groups=continuous_groups,
+            none_continuous_groups=none_continuous_groups,
+            is_fixed_mask=is_fixed_mask, is_fixed_key=is_fixed_key,
+            is_flexi_field=is_flexi_field, is_fixed_field=is_fixed_field,
+            shared_key_groups=shared_key_groups,
+            is_shared_key=is_shared_key, is_continuous=is_continuous)
+
+        # if not in the correct group, switch groups.
+        if group_currently_resides_in != real_group:
+
+            # create new tracked_set with this partition in front, as it
+            #  governed the move
+            new_tracked_set = OrderedSet()
+            new_tracked_set.add(partition)
+            for partition_to_move in tracked_set:
+                new_tracked_set.add(partition_to_move)
+
+            # remove mapping from wrong set, add to right set
+            group_currently_resides_in.remove(tracked_set)
+
+            # set tracked set to be new tracked set
+            tracked_set = new_tracked_set
+            real_group.append(tracked_set)
+
+            # update partition mapping accordingly
+            for partition in tracked_set:
+                partition_mappings[partition] = (tracked_set, real_group)
+
+            # verify the move is a valid move (no conflicting constraints)
+            _verify_constraints(
+                tracked_set=tracked_set, fixed_mask_groups=fixed_mask_groups,
+                continious_key_groups=continuous_groups,
+                fixed_field_groups=fixed_field_groups,
+                fixed_key_groups=fixed_key_groups,
+                group=group_currently_resides_in,
+                flexi_field_groups=flexi_field_groups,
+                share_key_groups=shared_key_groups)
+
+        # if share key, locate and find partitions to merge
+        if is_shared_key:
+            # get share key constraint that must be there
+            other_partitions = \
+                utility_calls.locate_first_constraint_of_type(
+                    partition.constraints, ShareKeyConstraint).other_partitions
+
+            # search for merging groups
+            tracked_set, real_group = _search_for_merger(
```

Comment:
Again something that would be avoided by first grouping the partitions by shared key constraint

Reply:
but that grouping is a massive search, what if the next partition in the list has a seperate share key set which includes more and then repeat over and over. And now add if each of those partitions also include other constraints, which make it no longer a share key on tis own but say a fixed key share key set.

Reply:
You must already be doing this though to get the grouping - I will have a look and see what can be done...

Reply:
i do it piece meal, handling each partition as it comes. allowing us 1 pass.